File: src/client/BackendMcpClient.ts
import type { ServerRegistration, ServerHealth } from '../types/server.ts';
import type {
  McpListPromptsResponse,
  McpListResourcesResponse,
  McpListToolsResponse,
  McpPromptGetRequest,
  McpPromptGetResponse,
  McpResourceReadRequest,
  McpResourceReadResponse,
  McpToolCallRequest,
  McpToolCallResponse,
} from '../types/mcp.ts';
import type { RoutingConfig } from '../types/config.ts';
import { HealthStatus } from '../types/server.ts';

/**
 * HTTP client for communicating with backend MCP servers
 */
export class BackendMcpClient {
  constructor(private config: RoutingConfig) {}

  /**
   * Call a tool on a backend server with retry logic
   */
  async callTool(
    server: ServerRegistration,
    toolName: string,
    args?: Record<string, unknown>
  ): Promise<McpToolCallResponse> {
    const request: McpToolCallRequest = {
      name: toolName,
      arguments: args,
    };

    return await this.retryRequest(async () => {
      const response = await fetch(`${server.endpoint}/tools/call`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(request),
        signal: AbortSignal.timeout(this.config.timeout.read),
      });

      if (!response.ok) {
        throw new Error(
          `Tool call failed: ${response.status} ${response.statusText}`
        );
      }

      return await response.json();
    });
  }

  /**
   * Read a resource from a backend server
   */
  async readResource(
    server: ServerRegistration,
    uri: string
  ): Promise<McpResourceReadResponse> {
    const request: McpResourceReadRequest = { uri };

    return await this.retryRequest(async () => {
      const response = await fetch(`${server.endpoint}/resources/read`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(request),
        signal: AbortSignal.timeout(this.config.timeout.read),
      });

      if (!response.ok) {
        throw new Error(
          `Resource read failed: ${response.status} ${response.statusText}`
        );
      }

      return await response.json();
    });
  }

  /**
   * Get a prompt from a backend server
   */
  async getPrompt(
    server: ServerRegistration,
    promptName: string,
    args?: Record<string, unknown>
  ): Promise<McpPromptGetResponse> {
    const request: McpPromptGetRequest = {
      name: promptName,
      arguments: args,
    };

    return await this.retryRequest(async () => {
      const response = await fetch(`${server.endpoint}/prompts/get`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(request),
        signal: AbortSignal.timeout(this.config.timeout.read),
      });

      if (!response.ok) {
        throw new Error(
          `Prompt get failed: ${response.status} ${response.statusText}`
        );
      }

      return await response.json();
    });
  }

  /**
   * List tools from a backend server
   */
  async listTools(server: ServerRegistration): Promise<McpListToolsResponse> {
    const response = await fetch(`${server.endpoint}/tools/list`, {
      method: 'GET',
      signal: AbortSignal.timeout(this.config.timeout.read),
    });

    if (!response.ok) {
      throw new Error(
        `List tools failed: ${response.status} ${response.statusText}`
      );
    }

    return await response.json();
  }

  /**
   * List resources from a backend server
   */
  async listResources(
    server: ServerRegistration
  ): Promise<McpListResourcesResponse> {
    const response = await fetch(`${server.endpoint}/resources/list`, {
      method: 'GET',
      signal: AbortSignal.timeout(this.config.timeout.read),
    });

    if (!response.ok) {
      throw new Error(
        `List resources failed: ${response.status} ${response.statusText}`
      );
    }

    return await response.json();
  }

  /**
   * List prompts from a backend server
   */
  async listPrompts(
    server: ServerRegistration
  ): Promise<McpListPromptsResponse> {
    const response = await fetch(`${server.endpoint}/prompts/list`, {
      method: 'GET',
      signal: AbortSignal.timeout(this.config.timeout.read),
    });

    if (!response.ok) {
      throw new Error(
        `List prompts failed: ${response.status} ${response.statusText}`
      );
    }

    return await response.json();
  }

  /**
   * Check health of a backend server
   */
  async checkHealth(server: ServerRegistration): Promise<ServerHealth> {
    const startTime = Date.now();

    try {
      const response = await fetch(
        `${server.endpoint.replace('/mcp', '')}/actuator/health`,
        {
          method: 'GET',
          signal: AbortSignal.timeout(this.config.timeout.connect),
        }
      );

      const latency = Date.now() - startTime;

      if (response.ok) {
        return {
          status: HealthStatus.HEALTHY,
          lastCheck: new Date(),
          latency,
          consecutiveFailures: 0,
        };
      } else {
        return {
          status: HealthStatus.DEGRADED,
          lastCheck: new Date(),
          latency,
          errorMessage: `HTTP ${response.status}`,
          consecutiveFailures: server.health.consecutiveFailures + 1,
        };
      }
    } catch (error) {
      const latency = Date.now() - startTime;
      return {
        status: HealthStatus.DOWN,
        lastCheck: new Date(),
        latency,
        errorMessage: error instanceof Error ? error.message : 'Unknown error',
        consecutiveFailures: server.health.consecutiveFailures + 1,
      };
    }
  }

  /**
   * Retry a request with exponential backoff
   */
  private async retryRequest<T>(fn: () => Promise<T>): Promise<T> {
    let lastError: Error | null = null;

    for (let attempt = 0; attempt < this.config.retry.maxAttempts; attempt++) {
      try {
        return await fn();
      } catch (error) {
        lastError = error instanceof Error ? error : new Error(String(error));

        // Don't retry on last attempt
        if (attempt === this.config.retry.maxAttempts - 1) {
          break;
        }

        // Calculate backoff delay
        const delay = Math.min(
          this.config.retry.backoffDelay *
            Math.pow(this.config.retry.backoffMultiplier, attempt),
          this.config.retry.maxDelay
        );

        console.warn(
          `Request failed (attempt ${attempt + 1}), retrying in ${delay}ms:`,
          lastError.message
        );
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
    }

    throw lastError || new Error('Request failed after retries');
  }
}

File: src/main/java/ch/sbb/mcp/gateway/client/BackendMcpClient.java
package ch.sbb.mcp.gateway.client;

import ch.sbb.mcp.gateway.model.ServerHealth;
import ch.sbb.mcp.gateway.model.ServerRegistration;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.ResponseEntity;
import org.springframework.retry.support.RetryTemplate;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import java.time.Duration;
import java.time.Instant;
import java.util.Map;

/**
 * HTTP client for communicating with backend MCP servers.
 * 
 * <p>Provides methods for calling tools, reading resources, getting prompts,
 * and checking server health. All operations include retry logic.</p>
 */
@Service
public class BackendMcpClient {
    
    private static final Logger log = LoggerFactory.getLogger(BackendMcpClient.class);
    
    private final RestTemplate restTemplate;
    private final RetryTemplate retryTemplate;
    
    public BackendMcpClient(RestTemplate restTemplate, RetryTemplate retryTemplate) {
        this.restTemplate = restTemplate;
        this.retryTemplate = retryTemplate;
    }
    
    /**
     * Call a tool on a backend server.
     * 
     * @param server the server registration
     * @param toolName the tool name (without namespace)
     * @param arguments the tool arguments
     * @return the tool result
     */
    public Map<String, Object> callTool(ServerRegistration server, String toolName, Map<String, Object> arguments) {
        return retryTemplate.execute(context -> {
            String url = server.endpoint() + "/tools/call";
            
            Map<String, Object> request = Map.of(
                "name", toolName,
                "arguments", arguments
            );
            
            log.debug("Calling tool {} on server {} at {}", toolName, server.id(), url);
            
            ResponseEntity<Map> response = restTemplate.postForEntity(url, request, Map.class);
            
            @SuppressWarnings("unchecked")
            Map<String, Object> result = response.getBody();
            
            return result;
        });
    }
    
    /**
     * Read a resource from a backend server.
     * 
     * @param server the server registration
     * @param uri the resource URI
     * @return the resource content
     */
    public String readResource(ServerRegistration server, String uri) {
        return retryTemplate.execute(context -> {
            String url = server.endpoint() + "/resources/read";
            
            Map<String, Object> request = Map.of("uri", uri);
            
            log.debug("Reading resource {} from server {} at {}", uri, server.id(), url);
            
            ResponseEntity<String> response = restTemplate.postForEntity(url, request, String.class);
            
            return response.getBody();
        });
    }
    
    /**
     * Get a prompt from a backend server.
     * 
     * @param server the server registration
     * @param promptName the prompt name (without namespace)
     * @param arguments the prompt arguments
     * @return the prompt result
     */
    public Map<String, Object> getPrompt(ServerRegistration server, String promptName, Map<String, Object> arguments) {
        return retryTemplate.execute(context -> {
            String url = server.endpoint() + "/prompts/get";
            
            Map<String, Object> request = Map.of(
                "name", promptName,
                "arguments", arguments != null ? arguments : Map.of()
            );
            
            log.debug("Getting prompt {} from server {} at {}", promptName, server.id(), url);
            
            ResponseEntity<Map> response = restTemplate.postForEntity(url, request, Map.class);
            
            @SuppressWarnings("unchecked")
            Map<String, Object> result = response.getBody();
            
            return result;
        });
    }
    
    /**
     * Check the health of a backend server.
     * 
     * @param server the server registration
     * @return the health status
     */
    public ServerHealth checkHealth(ServerRegistration server) {
        try {
            Instant start = Instant.now();
            String url = server.endpoint() + "/actuator/health";
            
            log.debug("Checking health of server {} at {}", server.id(), url);
            
            @SuppressWarnings("rawtypes")
            ResponseEntity<Map> response = restTemplate.getForEntity(url, Map.class);
            Duration latency = Duration.between(start, Instant.now());
            
            boolean isHealthy = response.getStatusCode().is2xxSuccessful();
            
            return ServerHealth.builder()
                .status(isHealthy ? ServerHealth.HealthStatus.HEALTHY : ServerHealth.HealthStatus.DEGRADED)
                .lastCheck(Instant.now())
                .latency(latency)
                .consecutiveFailures(0)
                .build();
                
        } catch (Exception e) {
            log.warn("Health check failed for server {}: {}", server.id(), e.getMessage());
            
            int consecutiveFailures = server.health().consecutiveFailures() + 1;
            
            return ServerHealth.builder()
                .status(ServerHealth.HealthStatus.DOWN)
                .lastCheck(Instant.now())
                .errorMessage(e.getMessage())
                .consecutiveFailures(consecutiveFailures)
                .build();
        }
    }
    
    /**
     * List tools from a backend server.
     * 
     * @param server the server registration
     * @return list of tools
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> listTools(ServerRegistration server) {
        try {
            String url = server.endpoint() + "/tools/list";
            
            log.debug("Listing tools from server {} at {}", server.id(), url);
            
            @SuppressWarnings("rawtypes")
            ResponseEntity<Map> response = restTemplate.postForEntity(url, Map.of(), Map.class);
            
            return response.getBody();
        } catch (Exception e) {
            log.error("Failed to list tools from server {}: {}", server.id(), e.getMessage());
            return Map.of("tools", java.util.List.of());
        }
    }
    
    /**
     * List resources from a backend server.
     * 
     * @param server the server registration
     * @return list of resources
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> listResources(ServerRegistration server) {
        try {
            String url = server.endpoint() + "/resources/list";
            
            log.debug("Listing resources from server {} at {}", server.id(), url);
            
            @SuppressWarnings("rawtypes")
            ResponseEntity<Map> response = restTemplate.postForEntity(url, Map.of(), Map.class);
            
            return response.getBody();
        } catch (Exception e) {
            log.error("Failed to list resources from server {}: {}", server.id(), e.getMessage());
            return Map.of("resources", java.util.List.of());
        }
    }
    
    /**
     * List prompts from a backend server.
     * 
     * @param server the server registration
     * @return list of prompts
     */
    @SuppressWarnings("unchecked")
    public Map<String, Object> listPrompts(ServerRegistration server) {
        try {
            String url = server.endpoint() + "/prompts/list";
            
            log.debug("Listing prompts from server {} at {}", server.id(), url);
            
            @SuppressWarnings("rawtypes")
            ResponseEntity<Map> response = restTemplate.postForEntity(url, Map.of(), Map.class);
            
            return response.getBody();
        } catch (Exception e) {
            log.error("Failed to list prompts from server {}: {}", server.id(), e.getMessage());
            return Map.of("prompts", java.util.List.of());
        }
    }
}
